src/librustc/ty/context.rs

```diff
@@ -1756,6 +1740,16 @@ impl<'a, 'tcx> Lift<'tcx> for &'a Const<'a> {
     }
 }
 
+impl<'a, 'tcx> Lift<'tcx> for &'a mir::interpret::Allocation {
```

Comment:
Why is that needed now but was not needed before?

Reply:
It's required for https://github.com/rust-lang/rust/pull/56723/files

which is required for

https://github.com/rust-lang/rust/pull/56723/files#diff-e826542d6f24137d5b0b5ae333230529R1240

which now leads to https://github.com/rust-lang/rust/pull/56723/files#diff-ce39564003916a7d91474a3f63b5fee7R503

which doesn't fail to lift if lifting isn't zero cost like the old code did

https://github.com/rust-lang/rust/pull/56723/files#diff-c8a6d543f758cb294320bcac3b5268aeL1754

Reply:
This all happened after changing `&'tcx Const<'tcx>` to `Const<'tcx>`, because we don't need to intern before using such values everywhere. Without being interned, lifting is not a nop, thus causing lifting to fail.

Reply:
Your first two links go nowhere here, so I'm afraid I cannot follow.

So, normal lifting assumes that the data is already lifted, and just does a lookup in the `tcx.interners`?  (However that works, something has got to do the actual lifting after all.)

But now somehow this doesn't happen any more, data is not already lifted -- and also somehow allocations are *not* magically already lifted when we get here so we actually need to do work?  I am still confused, no idea what is going on here.

Also there's not even a dozen lifting instances here, are these really all types that ever need lifting?

Reply:
> Your first two links go nowhere here, so I'm afraid I cannot follow.

hmm weird, the links might point to code that only shows when clicking the expand buttons on github, and github doesn't auto-expand when given a link.

> So, normal lifting assumes that the data is already lifted, and just does a lookup in the tcx.interners? (However that works, something has got to do the actual lifting after all.)

Lifting for the types which have `Lift` impls in `context.rs`, yes. This works by assuming that some other code already interned the element in the global allocator. It wasn't lifted to be in the global allocator, it was just interned there. It might even be that the current instance we have just has a `'tcx` lifetime and not a `'gcx` lifetime due to some casting/coercion, even if it was allocated in the global allocator and does indeed live for the `'gcx` lifetime.

> But now somehow this doesn't happen any more, data is not already lifted -- and also somehow allocations are not magically already lifted when we get here so we actually need to do work?

We used to intern `ty::Const` for the return value of `const_eval` (this is how we got a `&'tcx ty::Const`). Since `const_eval` now returns `ty::Const`, we do not intern this value anymore. So, when lifting a value that contains a `ty::Const` (e.g. `ty::LazyConst`), we end up trying to lift the `Allocation` inside

> Also there's not even a dozen lifting instances here, are these really all types that ever need lifting?

No, the other instances are defined in different files. All these impls are dirstributed around `librustc`.